### PR TITLE
Create Pester Tests for Test-SecretVault in KeePass extension

### PR DESCRIPTION
**Pull request by trir262**
Hello Justin,
This PR adds some tests for the Test-SecretVault function in your extension module.
Can you take a look at it and if you feel this is OK, let me know and I'll try to add more tests for your other functions as well.

Summary:
* Created the tests in a tests folder in the extensions module, rather than in the main module.
* Used 'InModuleScope' to be able to test on script data & variables
* added small kdbx database to use as a base for all following tests

### CODE REVIEW

**Comment by JustinGrote** [comment]
Thanks these are great! I was going to create a matrix of tests for master password, key file, and windows account, this is a great start! I'll review more tomorrow.

**Comment by trir262** [comment]
My 2 cents: 
Since you have 2 powershell modules (of which one is an extension of the other), I think you will need 2 test scenario's:
* Tests for 'your' individual functions in your extension (like I have created now)
* Tests to verify if your entire module behaves like a 'SecretManagement' module (like: does it throw when the entry does not exist & so)

File: SecretManagement.KeePass.Extension/Tests/Test-SecretVault.Tests.ps1
Comment (JustinGrote) [inline SecretManagement.KeePass.Extension/Tests/Test-SecretVault.Tests.ps1:2]: Move this to the top as #requires -module Microsoft.Powershell.SecretManagement

File: SecretManagement.KeePass.Extension/Tests/Test-SecretVault.Tests.ps1
Comment (JustinGrote) [inline SecretManagement.KeePass.Extension/Tests/Test-SecretVault.Tests.ps1:1]: Needs to go to beforeEach, in Pester 5 these will run during test discovery which is not what we want.

File: SecretManagement.KeePass.Extension/Tests/Test-SecretVault.Tests.ps1
Comment (JustinGrote) [inline SecretManagement.KeePass.Extension/Tests/Test-SecretVault.Tests.ps1:3]: Needs to go to beforeEach, in Pester 5 these will run during test discovery which is not what we want.

File: SecretManagement.KeePass.Extension/Tests/Test-SecretVault.Tests.ps1
Comment (JustinGrote) [inline SecretManagement.KeePass.Extension/Tests/Test-SecretVault.Tests.ps1:4]: Needs to go to beforeEach, in Pester 5 these will run during test discovery which is not what we want.

File: SecretManagement.KeePass.Extension/Tests/Test-SecretVault.Tests.ps1
Comment (JustinGrote) [inline SecretManagement.KeePass.Extension/Tests/Test-SecretVault.Tests.ps1:6]: InModuleScope only needs to wrap the It statements that run in the module context.

File: SecretManagement.KeePass.Extension/Tests/Test-SecretVault.Tests.ps1
Comment (JustinGrote) [inline SecretManagement.KeePass.Extension/Tests/Test-SecretVault.Tests.ps1:48]: I don't understand this test. You say "should not request" but your assertion is that it called Get-Credential meaning it did prompt.

**Comment by JustinGrote** [comment]
@trir262 did you write these tests before I did my rewrite? I think I cover a lot of these now. Great work though, and some general feedback for cleanup. The github actions runner is also in place so you can verify the tests pass on every commit.

File: SecretManagement.KeePass.Extension/Tests/Test-SecretVault.Tests.ps1
Comment (trir262) [reply SecretManagement.KeePass.Extension/Tests/Test-SecretVault.Tests.ps1:48]: On the first pass of 'Test-Secret', the amount of times the cmdlet 'Get-Credential' is executed (mocked in the tests) is 1 because you want to request a credential.
From that point on, the 'Get-Credential' is no longer invoked anytime you run Test-SecretVault because your private variable Vault_.... alreay exists.
This tests describes exactly that: the number of times 'Get-Credential' is executed within the Context scope remains at 1

**Comment by trir262** [comment]
@JustinGrote did you merge the PR or not yet? Do you want me to update tests so the CI's you set pass?

**Comment by JustinGrote** [comment]
Ready for you to fix your tests so they "go green" on the checks here. Right now it looks like its not loading the secretmanagement.keepass correctly on your tests
image

File: SecretManagement.KeePass.Extension/Tests/Test-SecretVault.Tests.ps1
Comment (JustinGrote) [reply SecretManagement.KeePass.Extension/Tests/Test-SecretVault.Tests.ps1:48]: OK, I think Assert-MockCalled is scoped within the It though isn't it? meaning that it'll say 1 because get-credential was called. I may be wrong there.

**Comment by trir262** [comment]
Is it the idea that both test files go green prior to continuing?

**Comment by JustinGrote** [comment]
Correct, as of my last commit all my tests pass
https://github.com/JustinGrote/SecretManagement.KeePass/runs/1689097146

So I expect yours to all pass before I merge this in (and in fact merging is blocked until they all pass thanks to the Github Action I added)

File: .github/workflows/Invoke-Pester.ps1
Comment (JustinGrote) [inline .github/workflows/Invoke-Pester.ps1:2]: Please remove this, you should do this work in your tests, this broke my tests by doing it.

Comment (trir262) [reply .github/workflows/Invoke-Pester.ps1]: will do, thanks

**Comment by trir262** [comment]
@JustinGrote I think I got it now...

I'll continue working on other scenario's for the 'Test-SecretVault' so that other authentication methods are tested as well.

File: SecretManagement.KeePass.Extension/Tests/Test-SecretVault.Tests.ps1
Comment (JustinGrote) [inline SecretManagement.KeePass.Extension/Tests/Test-SecretVault.Tests.ps1:83]: nit: Pipe this to Out-Null since you dont actually use TheVault to avoid a warning here

File: SecretManagement.KeePass.Extension/Tests/Test-SecretVault.Tests.ps1
Comment (JustinGrote) [reply SecretManagement.KeePass.Extension/Tests/Test-SecretVault.Tests.ps1:55]: Scope context makes this work as expected, thanks. However this test is not individually validatable, it requires the context of another test or it will fail if tested individually. I'll fix it later if it really annoys me :)

File: SecretManagement.KeePass.Extension/Tests/Test-SecretVault.Tests.ps1
Comment (JustinGrote) [reply SecretManagement.KeePass.Extension/Tests/Test-SecretVault.Tests.ps1:83]: @trir262 once this little nit is fixed (its an issue in two places) then I'll merge it.

**Comment by JustinGrote** [comment]
@trir262 just request a review when youre finished. Thanks for the contribution!

File: SecretManagement.KeePass.Extension/Tests/Test-SecretVault.Tests.ps1
Comment (trir262) [reply SecretManagement.KeePass.Extension/Tests/Test-SecretVault.Tests.ps1:52]: I added it because it is a real life scenario:  you don't want to make an error in the Test-SecretVault function that will constantly request the credential information over and over again.

**Comment by trir262** [comment]
@JustinGrote I added some function parameter tests and I think I'm done.
Can you give it a last review and if ok complete the PR?

File: SecretManagement.KeePass.Extension/Tests/Test-SecretVault.Tests.ps1
Comment (JustinGrote) [reply SecretManagement.KeePass.Extension/Tests/Test-SecretVault.Tests.ps1:81]: Right, it makes sense as a test, however it should be structured all within a single "it" block. If I click "run test" on that test without running it on another it will fail because it requires previous state that isn't captured in a beforeeach/beforeall or mocked.

**Review by JustinGrote — APPROVED** [review]

**Pull request merged**
